In [1]:
# TEXT EMBEDDINGS PROJECT

import pandas as pd
import numpy as np
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

print("Libraries imported successfully!")

c:\Users\AFRA HANEEN\AppData\Local\Programs\Python\Python314\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Libraries imported successfully!


In [2]:
# Load the sentence dataset

df = pd.read_csv("Data/text_generation_results.csv")

print("Dataset loaded successfully!")
print("Number of sentences:", len(df))

df.head()

Dataset loaded successfully!
Number of sentences: 20


,id,sentence,topic
0,1,Artificial intelligence is transforming many i...,AI
1,2,Machine learning enables computers to learn pa...,AI
2,3,Deep learning uses neural networks to solve co...,AI
3,4,Students can improve their academic performanc...,Education
4,5,Online education allows students to access lea...,Education


In [3]:
# Get sentences from the dataset

sentences = df["sentence"].tolist()

print("Sentences extracted successfully!")
print("Total sentences:", len(sentences))

Sentences extracted successfully!
Total sentences: 20


In [4]:
# Load the embedding model

model = SentenceTransformer("all-MiniLM-L6-v2")

print("Embedding model loaded successfully!")


Loading weights: 100%|██████████| 103/103 [00:00<00:00, 3360.79it/s]


Embedding model loaded successfully!


In [5]:
# Generate embeddings

embeddings = model.encode(sentences)

print("Embeddings generated successfully!")
print("Embedding shape:", embeddings.shape)

Embeddings generated successfully!
Embedding shape: (20, 384)


In [6]:
# Save embeddings

np.save("Data/embeddings.npy", embeddings)

print("Embeddings saved successfully!")

Embeddings saved successfully!


In [7]:
# Calculate cosine similarity

similarity_matrix = cosine_similarity(embeddings)

print("Cosine similarity calculated successfully!")
print("Similarity matrix shape:", similarity_matrix.shape)

Cosine similarity calculated successfully!
Similarity matrix shape: (20, 20)


In [8]:
# Create sentence pairs with similarity scores

results = []

for i in range(len(sentences)):
    for j in range(i + 1, len(sentences)):
        results.append({
            "Text 1": sentences[i],
            "Text 2": sentences[j],
            "Cosine Similarity": round(similarity_matrix[i][j], 4)
        })

results_df = pd.DataFrame(results)

print("Sentence pairs created:", len(results_df))
results_df.head()

Sentence pairs created: 190


,Text 1,Text 2,Cosine Similarity
0,Artificial intelligence is transforming many i...,Machine learning enables computers to learn pa...,0.3963
1,Artificial intelligence is transforming many i...,Deep learning uses neural networks to solve co...,0.4577
2,Artificial intelligence is transforming many i...,Students can improve their academic performanc...,0.0873
3,Artificial intelligence is transforming many i...,Online education allows students to access lea...,0.0735
4,Artificial intelligence is transforming many i...,Regular exercise helps people maintain a healt...,0.0550


In [9]:
# Sort pairs by similarity

results_df = results_df.sort_values(
    by="Cosine Similarity",
    ascending=False
)

results_df.head(5)

,Text 1,Text 2,Cosine Similarity
162,Space exploration helps scientists understand ...,Scientists are developing new technologies to ...,0.5968
124,Reducing carbon emissions can help slow the ef...,Renewable energy sources such as solar and win...,0.5812
112,Climate change is causing significant changes ...,Reducing carbon emissions can help slow the ef...,0.5493
85,Regular exercise helps people maintain a healt...,Eating a balanced diet provides the body with ...,0.4988
145,Football is one of the most popular sports in ...,Cricket matches attract millions of viewers an...,0.4823


In [10]:
# Save all similarity results

results_df.to_csv(
    "Data/similarity_results.csv",
    index=False
)

print("Similarity results saved successfully!")

Similarity results saved successfully!


In [11]:
# Save top 10 most similar pairs

top_10 = results_df.head(10)

top_10.to_csv(
    "Data/top_10_pairs.csv",
    index=False
)

print("Top 10 pairs saved successfully!")

Top 10 pairs saved successfully!


In [12]:
# Display Top 5 Most Similar Sentence Pairs

print("TOP 5 MOST SIMILAR SENTENCE PAIRS\n")

top_5 = results_df.head(5)

for index, row in top_5.iterrows():
    print("Text 1:", row["Text 1"])
    print("Text 2:", row["Text 2"])
    print("Cosine Similarity:", row["Cosine Similarity"])
    print("-" * 80)

TOP 5 MOST SIMILAR SENTENCE PAIRS

Text 1: Space exploration helps scientists understand planets, stars, and the universe.
Text 2: Scientists are developing new technologies to explore distant planets and moons.
Cosine Similarity: 0.5968000292778015
--------------------------------------------------------------------------------
Text 1: Reducing carbon emissions can help slow the effects of climate change.
Text 2: Renewable energy sources such as solar and wind power can reduce dependence on fossil fuels.
Cosine Similarity: 0.5812000036239624
--------------------------------------------------------------------------------
Text 1: Climate change is causing significant changes in global weather patterns.
Text 2: Reducing carbon emissions can help slow the effects of climate change.
Cosine Similarity: 0.5493000149726868
--------------------------------------------------------------------------------
Text 1: Regular exercise helps people maintain a healthy and active lifestyle.
Text 2: Eat

In [13]:
# Final Project Summary

print("TEXT EMBEDDINGS PROJECT")
print("-----------------------")
print("Total sentences:", len(sentences))
print("Embedding dimensions:", embeddings.shape[1])
print("Total sentence pairs compared:", len(results_df))
print("Top 5 pairs identified: Yes")
print("Embeddings saved: Yes")
print("Similarity results saved: Yes")
print("Project completed successfully!")

TEXT EMBEDDINGS PROJECT
-----------------------
Total sentences: 20
Embedding dimensions: 384
Total sentence pairs compared: 190
Top 5 pairs identified: Yes
Embeddings saved: Yes
Similarity results saved: Yes
Project completed successfully!


# TEXT EMBEDDINGS PROJECT

## Objective
To generate text embeddings for 20 sentences and compare their semantic similarity using cosine similarity.

## Methodology

1. Collected 20 sentences from different topics.
2. Loaded the Sentence Transformer model.
3. Converted the sentences into 384-dimensional embeddings.
4. Calculated cosine similarity between sentence pairs.
5. Compared 190 unique sentence pairs.
6. Identified the top 5 most similar pairs.
7. Saved embeddings and similarity results as output files.

## Results

The embeddings were generated successfully for all 20 sentences. A total of 190 unique sentence pairs were compared using cosine similarity.

### Top 5 Most Similar Pairs

The top five pairs were identified based on their cosine similarity scores. Higher scores indicate greater semantic similarity.

## Observations

- Sentence embeddings represent the semantic meaning of text as numerical vectors.
- Related topics generally produced higher cosine similarity scores.
- Space-related and environment-related sentences showed relatively high similarity.
- Unrelated topics generally produced lower similarity scores.
- Embeddings can identify meaning even when the exact words are different.

## Conclusion

This project demonstrates how text embeddings can be used to represent sentence meaning and measure semantic similarity using cosine similarity. Text embeddings are useful in semantic search, recommendation systems, document retrieval and LLM applications.